autoextract_spiders/middlewares.py

```diff
@@ -0,0 +1,91 @@
+
+Usage:
+
+SPIDER_MIDDLEWARES = {
+    'autoextract_spiders.middlewares.S3ItemWriteMiddleware': 1,
+}
+
+To enable the middleware, set the target bucket:
+
+S3_WRITE_BUCKET = <bucket name>
+
+Spider that needs to write items to s3 must set following attributes:
+
+s3_write_folder - Folder where to write the output
+s3_write_item_type - Only items belonging to the given class are written to s3
+
+AWS credentials are taken from the environment variables AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY.
+"""
+
+import os
+import json
+import logging
+from hashlib import md5
+from datetime import datetime
+
+from pytz import utc, timezone
+from s3fs import S3FileSystem
+
+from scrapy import signals
+from scrapy.exceptions import NotConfigured
+from scrapy.exporters import PythonItemExporter
+
+TIMEZONE = timezone('Europe/Stockholm')
+
+
+class S3ItemWriteMiddleware:
+
+    def __init__(self, crawler):
+        self.bucket = crawler.settings.get('S3_WRITE_BUCKET')
+        if not self.bucket:
+            raise NotConfigured
+        self.log = logging.getLogger(__name__)
+        self.s3 = S3FileSystem(key=os.environ.get('AWS_ACCESS_KEY_ID'),
+                               secret=os.environ.get('AWS_SECRET_ACCESS_KEY'))
+        self.s3folder = None
+        self.item_type = None
+        self.enabled = False
+        self.exporter = PythonItemExporter(binary=False)
+
+    @classmethod
+    def from_crawler(cls, crawler):
+        s = cls(crawler)
+        crawler.signals.connect(s.spider_opened, signal=signals.spider_opened)
+        return s
+
+    def process_spider_output(self, response, result, spider):
+        for item in result:
+            if self.enabled and isinstance(item, self.item_type):
+                self.write_item(item)
+            yield item
+
+    def spider_opened(self, spider):
+        s3folder = getattr(spider, 's3_write_folder', None)
+        if s3folder:
+            self.s3folder = os.path.join(self.bucket, s3folder)
+        self.item_type = getattr(spider, 's3_write_item_type', None)
+        self.enabled = self.s3folder and self.item_type
+        if self.enabled:
+            self.log.info(f'S3ItemWriter will output items in "{self.s3folder}".')
+        else:
+            self.log.info(f'S3ItemWriter is disabled.')
+
+    def write_item(self, item):
+        fname = self.item_fname(item)
+        path = os.path.join(self.s3folder, fname)
+        with self.s3.open(path, 'wb') as f:
+            exported_item = self.exporter.export_item(item)
+            for key in list(exported_item.keys()):
+                if key.startswith('_'):
+                    exported_item.pop(key)
+            f.write(json.dumps(exported_item, indent=2, sort_keys=True).encode('utf8'))
+
+    @staticmethod
+    def item_fname(item):
+        mhash = md5()
+        mhash.update(item['url'].encode('utf8'))
+        date_now = datetime.now(utc).astimezone(TIMEZONE)
+        folder = date_now.strftime('%Y/%m/%d/%H/%M/%S')
```

Comment:
This folder structure should not be hardocoded either.
This structure will create /year/month/day/hour/minute/second/ which is not what a lot of users would expect.
This could also be a setting per project.